### PR TITLE
Add liveness tracking to LazyTensor

### PR DESCRIPTION
**Pull request by bgogul**
Tracks list of `LazyTensorOperation` instances held in `LazyTensor` instances. This is required to perform batch materialization of all live tensors during lazy evaluation.

### CODE REVIEW

File: Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift [hunk 1]
```diff
@@ -36,12 +36,72 @@ class LazyTensor: _AnyTensorHandle {
...
+        let op: LazyTensorOperation
+        let liveRefCount: Int
+        let allRefCount: Int
+    }
+
+    private static var operationRefCounts: [
+        ObjectIdentifier: LazyTensorOperationRefCounts] = [:]
+
+    static func incrementRefCount(_ op: LazyTensorOperation, isLive: Bool) {
+        let opId = ObjectIdentifier(op)
```
Comment (rxwei) [inline Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift]: `Id` -> `ID`

File: Tests/TensorFlowTests/LazyTensorTests.swift [hunk 2]
```diff
@@ -16,6 +16,18 @@ import XCTest
 @testable import TensorFlow
 import CTensorFlow
 
+struct LazyTensorOperationRef : Equatable, Hashable {
```
[the remark this answers is not in the document]
Comment (bgogul) [reply Tests/TensorFlowTests/LazyTensorTests.swift]: Done.

File: Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift [hunk 3]
```diff
@@ -36,12 +36,72 @@ class LazyTensor: _AnyTensorHandle {
...
+    static func isLive(_ op: LazyTensorOperation) -> Bool {
+        let opId = ObjectIdentifier(op)
+        if let counts = operationRefCounts[opId] {
+            return counts.liveRefCount > 0
+        }
+        return false
+    }
+
+    static func onLiveOperations(_ perform: (LazyTensorOperation) -> ()) {
+        for (_, counts) in operationRefCounts {
```
[the remark this answers is not in the document]
Comment (bgogul) [reply Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift]: Done.

File: Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift [hunk 4]
```diff
@@ -36,12 +36,72 @@ class LazyTensor: _AnyTensorHandle {
...
+                    liveRefCount: counts.liveRefCount - (isLive ? 1 : 0),
+                    allRefCount: counts.allRefCount - 1)
+            } else {
+                operationRefCounts.removeValue(forKey: opId)
+            }
+        }
+    }
+
+    static func isLive(_ op: LazyTensorOperation) -> Bool {
+        let opId = ObjectIdentifier(op)
```
[the remark this answers is not in the document]
Comment (bgogul) [reply Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift]: Done

File: Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift (hunk 1 above)
Comment (bgogul) [reply Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift]: Done.

File: Tests/TensorFlowTests/LazyTensorTests.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (bgogul) [reply Tests/TensorFlowTests/LazyTensorTests.swift]: Actually, I want that as well for some caching in lazy tensor implementation. However, not all properties of `LazyTensorOperation` will automatically conform to `Equatable` and `Hashable`. e.g., `Attribute` has some associated types that don't conform. 

I will send out a separate PR with all the needed changes.

File: Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (bgogul) [reply Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift]: Done.

File: Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift [hunk 5]
```diff
@@ -36,12 +36,72 @@ class LazyTensor: _AnyTensorHandle {
...
+        if let counts = operationRefCounts[opId] {
+            return counts.liveRefCount > 0
+        }
+        return false
+    }
+
+    static func onLiveOperations(_ perform: (LazyTensorOperation) -> ()) {
+        for (_, counts) in operationRefCounts {
+            if (counts.liveRefCount > 0) { perform(counts.op) }
```
[the remark this answers is not in the document]
Comment (bgogul) [reply Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift]: Done/

File: Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift [hunk 6]
```diff
@@ -36,12 +36,72 @@ class LazyTensor: _AnyTensorHandle {
...
+
+    // Liveness tracking for LazyTensorOperations
+    //
+    struct LazyTensorOperationRefCounts {
+        let op: LazyTensorOperation
+        let liveRefCount: Int
+        let allRefCount: Int
+    }
+
+    private static var operationRefCounts: [
```
[the remark this answers is not in the document]
Comment (bgogul) [reply Sources/TensorFlow/LazyTensor/LazyTensorOperation.swift]: Didn't realize that we can go up to 100 chars. Fixed now.

**Comment by bgogul** [comment]
@rxwei Thanks for the review. I addressed your comments. PTAL.

**Pull request merged**